text/0000-volta-use.md

```diff
@@ -0,0 +1,103 @@
+
+ARGS:
+    <tool[@version]>... Tools to make available, like `node@lts` or `yarn@^1.14`
+```
+
+## Behavior
+
+The behavior of `volta use` will ultimately be the same as the current `volta install` or `volta pin`. However, when run without flags, it will be contextual, based on where the command is being run. So if a user runs `volta use node@18` in a pinned project, the project will be updated to reflect the newly selected version. If they instead run `volta use node@18` _outside_ of a pinned project, then the default Node version will be updated. This ensures that running `volta use node@version` will always update the version that is currently active, meaning the next run of `node` will use that updated version.
+
+Notably, if a user runs `volta use node` within a project that doesn't currently have a pinned version (i.e. has no `"volta"` entry in `package.json`), then the command will update the global / default version of Node, it will _not_ create a pin if one doesn't exist.
+
+If a user wants to change a version that isn't currently active, they can use the flags to change the behavior.
+
+### `--global`
+
+Running `volta use --global` will work identically to the existing `volta install`, updating the default version regardless of whether the current directory is part of a pinned project.
+
+### `--pin`
+
+Similarly, `volta use --pin` will work identically to the current `volta pin`, always updating the current project. This includes creating a `volta` key in `package.json` if it isn't there already.
+
+## Implementation
+
+The technical implementation of `volta use` should be fairly straightforward. The new command will use the flags and context (i.e. if we are currently in a pinned project or did the user pass `--global`) to determine whether the behavior should be equivalent to `volta install` or `volta pin`. It will then delegate to the appropriate existing command behavior based on that determination. The ultimate flow should be:
+
+- If `--global` was set -> use `volta install` behavior
+- If `--pin` was set -> use `volta pin` behavior
+- If _both_ were set, error (this can be handled by Clap I believe by making the two flags exclusive)
+- If neither flag was set:
+  - If we are in a pinned project (one that already has a `"volta"` key in `package.json`) -> use `volta pin` behavior.
+  - Otherwise, use `volta install` behavior.
+
+# Critique
+
+[critique]: #critique
+
+## Confusion around omitting version
+
+One nice part about the existing `volta install` and `volta pin` commands is that they naturally work well with choosing a tool without selecting a version. For example, running `volta install node` makes intuitive sense as "pick whatever version and make it available." By contrast, `volta use node` doesn't immediately (to me, at least!) jump out as _picking_ a version of Node. `volta use node@lts` or similar, _do_ make sense, but omitting the version feels a little more awkward with `use` as the verb.
```

Comment:
One alternative this suggests is to actually *not* change `pin` to `use`, and just to stick with `pin`, but add the `--default` flag to it. That eliminates this oddity (which is, you are right, fairly significant!), and also gets rid of the need to introduce a new verb.

One thing that makes me think of: given that part of the motivation is that `pin` does not *necessarily* imply fetching, and that we *do* have a `fetch` command, what if we added a `--no-fetch` flag to `pin` as well? That feels to me like a really clear mental model:

> `pin` sets the version you ask for, and as a convenience, also fetches it if you don’t have it. If you want to `pin` a version *without* fetching it, you can run `pin --no-fetch`.

One nice thing that could unlock is the ability to set a version in a nice way while working offline. That would make it easy to use `volta pin` in e.g. project boilerplate generation without a network connection, for example. I have no idea if anyone wants that latter capability, but having it work offline, e.g. on a plane, even if that means I won’t be able to run using that, seems nice in some ways.

Reply:
I still need to let it rattle around my brain a bit, but the more I think about it, the more I'm coming around to the idea of `volta pin --default`. I do really like the verb making it clear what's happening, and adding a `--no-fetch` option would further cement the mental model.

Reply:
It's slow-moving, but letting this ruminate on some of my recent runs, I really like keeping `pin` as the verb. It already clearly indicates the important part of the operation: pinning the version to use. When I get some free time during the long weekend, I'm planning to update this RFC to reflect that approach.

The main sticking point I have is that I think we want a good "short" flag for pinning the default version to make it short to type out, especially for the "Getting Started" guide. Using `--default` as the long flag doesn't work great, because the short version `-d` intuitively makes me think of a "delete" operation, rather than a "default".

While I know we've generally used `default` instead of `global`, I wonder if using `--global` here would make sense. It would align with people's experiences with package managers to make it clear what `volta pin -g node` means. Also, part of Volta's value prop is that it makes "globals" like Node, npm, and 3rd-party packages safe to use by delegating to the local version when it's available.